### PR TITLE
Finish AuthBDD cases, address the fallout

### PR DESCRIPTION
**Pull request by coveros-matt-taylor**

### CODE REVIEW

File: src/main/kotlin/coverosR3z/logging/Logger.kt [hunk 1]
```diff
@@ -5,5 +5,7 @@ import coverosR3z.authentication.CurrentUserAccessor
 val cua = CurrentUserAccessor()
 
 fun logInfo(msg : String) {
-    println("INFO: ${cua.get().name}: $msg")
+    try {
```
Comment (7ep) [inline src/main/kotlin/coverosR3z/logging/Logger.kt]: I think there might be a more elegant way to handle this.

File: src/main/kotlin/coverosR3z/persistence/PureMemoryDatabase.kt [hunk 2]
```diff
@@ -48,7 +48,7 @@ class PureMemoryDatabase {
 
     fun addNewUser(userName: UserName, hash: Hash, salt: String) : Int {
         val newIndex = users.size + 1
-        users.add(User(newIndex, userName.value, hash, salt, null))
+        users.add(User(newIndex, userName.value, hash, salt, newIndex))
```
Comment (7ep) [inline src/main/kotlin/coverosR3z/persistence/PureMemoryDatabase.kt]: Debatable.  We should discuss this.

File: src/test/kotlin/coverosR3z/authentication/AuthenticationBDD.kt [hunk 3]
```diff
@@ -30,20 +27,43 @@ class AuthenticationBDD {
     @Test
     fun `I can add a time entry`() {
         // given I am logged in
+        val pmd = PureMemoryDatabase()
+        val cua = CurrentUserAccessor() // We need a real cua for this test, sharing a pmd with auth persistence
```
Comment (7ep) [inline src/test/kotlin/coverosR3z/authentication/AuthenticationBDD.kt]: Probably would be good to refactor this to its own method, and also this is a lot to read through, might benefit from comments and further refactoring.

File: src/test/kotlin/coverosR3z/timerecording/TimeRecordingTests.kt [hunk 4]
```diff
@@ -75,7 +77,7 @@ class TimeRecordingTests {
         val utils = TimeRecordingUtilities(fakeTimeEntryPersistence)
         val entry = createTimeEntryPreDatabase(time= Time(60 * 2), project= Project(1, "an invalid project"))
 
-        assertThrows(ExceededDailyHoursAmountException::class.java) { utils.recordTime(entry) }
+        assertThrows(AssertionError::class.java) { utils.recordTime(entry) }
```
Comment (7ep) [inline src/test/kotlin/coverosR3z/timerecording/TimeRecordingTests.kt]: Debatable - the name of the test method at the very least suggests further thought is warranted.

**Review by 7ep — COMMENTED** [review]
Coming along

File: src/main/kotlin/coverosR3z/logging/Logger.kt (hunk 1 above)
Comment (coveros-matt-taylor) [reply src/main/kotlin/coverosR3z/logging/Logger.kt]: Yeah, this was purely for getting through to the tests. Need to think about cua.get() failing more gracefully

File: src/main/kotlin/coverosR3z/persistence/PureMemoryDatabase.kt (hunk 2 above)
Comment (7ep) [reply src/main/kotlin/coverosR3z/persistence/PureMemoryDatabase.kt]: I think this requires more consensus on the design decision.

Comment (coveros-matt-taylor) [reply src/main/kotlin/coverosR3z/persistence/PureMemoryDatabase.kt]: Yeah, this is the simplest method of getting a unique id for them, which we need for this BDD case, but there are other options.

File: src/main/kotlin/coverosR3z/logging/Logger.kt (hunk 1 above)
Comment (7ep) [reply src/main/kotlin/coverosR3z/logging/Logger.kt]: There might be a way to segregate our logging - between that which is logging commands by a user versus being controlled by the system.  Or maybe when someone logs in the user is "system"?  Let's discuss.

File: src/test/kotlin/coverosR3z/timerecording/TimeRecordingTests.kt (hunk 4 above)
Comment (coveros-matt-taylor) [reply src/test/kotlin/coverosR3z/timerecording/TimeRecordingTests.kt]: I swear I changed this one back after I mocked cua appropriately, surprised it was passing

Comment (coveros-matt-taylor) [reply src/test/kotlin/coverosR3z/timerecording/TimeRecordingTests.kt]: It _is_ an assertion error thrown here

Comment (coveros-matt-taylor) [reply src/test/kotlin/coverosR3z/timerecording/TimeRecordingTests.kt]: Fixed using fakecua

File: src/main/kotlin/coverosR3z/persistence/PureMemoryDatabase.kt (hunk 2 above)
Comment (coveros-matt-taylor) [reply src/main/kotlin/coverosR3z/persistence/PureMemoryDatabase.kt]: Gone back to null

File: src/main/kotlin/coverosR3z/logging/Logger.kt (hunk 1 above)
Comment (coveros-matt-taylor) [reply src/main/kotlin/coverosR3z/logging/Logger.kt]: Easiest option: User-aware logging wrapping regular logging and prepending `$user:`
Second easiest: nullable logging/cua.get() call that quickly decides whether it has a user or whether it is `SYSTEM` or something of the like

File: src/test/kotlin/coverosR3z/authentication/AuthenticationBDD.kt (hunk 3 above)
Comment (coveros-matt-taylor) [reply src/test/kotlin/coverosR3z/authentication/AuthenticationBDD.kt]: The deed is done

**Comment by coveros-matt-taylor** [comment]
All reviews are resolved and all tests are back to passing. 

I've added a single test, ``Registering a user should require a employee identifier`()`, and made it pass with somewhat far-reaching but simple changes to `AuthenticationUtilities.register()` and the calls it makes all the way down to `PureMemoryDatabase`. 

The result is that we now require you to pass a (nullable) employeeId when registering a user. The idea being that a real-world user would be given some identifier by an administrator to use when registering themselves, which the system would use to link them to an employee pre-populated by the admin. It can, in the future, be more advanced than a simple integer, have password-level complexity, etc.

File: src/main/kotlin/coverosR3z/logging/Logger.kt (hunk 1 above)
Comment (7ep) [reply src/main/kotlin/coverosR3z/logging/Logger.kt]: Yeah, you mean, for easiest, something like this?
val user = CurrentUser.get() ?: "SYSTEM"

File: src/test/kotlin/coverosR3z/authentication/AuthenticationBDD.kt [hunk 5]
```diff
@@ -21,29 +18,71 @@ import coverosR3z.domainobjects.LoginStatuses as ls
...
      * people to carry out actions.  Like, Alice
      * can only modify Alice's stuff.
      */
+
+    fun `given I am logged in`(pmd: PureMemoryDatabase) {
+        cua.clearCurrentUserTestOnly() // We need to use cua for this test, sharing a pmd with auth persistence
+        // and time recording persistence, in order to avoid recordTime throwing a USER_EMPLOYEE_MISMATCH status
+        val authPersistence = AuthenticationPersistence(pmd)
+        val au = AuthenticationUtilities(authPersistence)
+        au.register(DEFAULT_USER.name, DEFAULT_PASSWORD, 1) // we know DEFAULT_EMPLOYEE has an id=1
```
Comment (7ep) [inline src/test/kotlin/coverosR3z/authentication/AuthenticationBDD.kt:34]: really nice and timely documentation

File: src/test/kotlin/coverosR3z/authentication/AuthenticationBDD.kt [hunk 6]
```diff
@@ -21,29 +18,71 @@ import coverosR3z.domainobjects.LoginStatuses as ls
...
+        au.login(DEFAULT_USER.name, DEFAULT_PASSWORD)
+
+        // Perform some quick checks
+        val user = authPersistence.getUser(UserName(DEFAULT_USER.name))
+
+        println("DEBUG")
```
Comment (7ep) [inline src/test/kotlin/coverosR3z/authentication/AuthenticationBDD.kt]: Do we still need these println's?

File: src/test/kotlin/coverosR3z/authentication/AuthenticationBDD.kt [hunk 7]
```diff
@@ -21,29 +18,71 @@ import coverosR3z.domainobjects.LoginStatuses as ls
...
+        println(cua.get())
+
+        assertEquals(user, cua.get()) // auth persistence and user persistence must agree
+        assertTrue("our user should be registered", au.isUserRegistered(DEFAULT_USER.name)) // registration must succeed
+    }
     @Test
     fun `I can add a time entry`() {
         // given I am logged in
+        val pmd = PureMemoryDatabase()
```
Comment (7ep) [inline src/test/kotlin/coverosR3z/authentication/AuthenticationBDD.kt:46]: In the BDD tests we want to refactor them so it's really easy to read each test and see what is the business intention.

File: src/test/kotlin/coverosR3z/authentication/AuthenticationBDD.kt [hunk 8]
```diff
@@ -53,34 +92,33 @@ class AuthenticationBDD {
         val au = AuthenticationUtilities(authPersistence)
 
         // when I register a new employee with employeename "matt" and password "asdfoiajwefowejf"
-        au.register("matt", "asdfoiajwefowejf")
+        au.register("matt", "asdfoiajwefowejf", 12, )
```
Comment (7ep) [inline src/test/kotlin/coverosR3z/authentication/AuthenticationBDD.kt]: comma at the end doesn't seem to provide any extra clarity to my eyes.

File: src/test/kotlin/coverosR3z/authentication/UserManagementTests.kt [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+import coverosR3z.createTimeEntryPreDatabase
+import org.junit.Assert.assertEquals
+
+class UserManagementTests {
+    /**
+     * When a user is registered, they must supply an employee identifier provided by some
+     * administrator. We cannot allow a user to simply create their corresponding employee--
+     * they must be invited into the org, with some prior setup.
+     */
+
```
Comment (7ep) [inline src/test/kotlin/coverosR3z/authentication/UserManagementTests.kt]: By convention, method comments shouldn't have whitespace between them and the method

**Review by 7ep — CHANGES_REQUESTED** [review]
A lot of good work here

File: src/test/kotlin/coverosR3z/authentication/AuthenticationBDD.kt (hunk 5 above)
Comment (coveros-matt-taylor) [reply src/test/kotlin/coverosR3z/authentication/AuthenticationBDD.kt:34]: To do `val user = CurrentUser.get() ?: "SYSTEM"` or something of the sort, we need to change this:
``` override fun get(): User {
        return CurrentUser.get()
                ?: throw AssertionError("Cannot record time when no user is logged in, you mangy cur" +
                        "Please be aware: if this occurred during a test, it is possible " +
                        "that it is due to parallel tests conflicting with each other.  In that " +
                        "case, now is the time to strongly consider having tests that use this run serially")
    }
```
and instead make get() nullable, which I am fine with

File: src/test/kotlin/coverosR3z/authentication/AuthenticationBDD.kt (hunk 6 above)
Comment (coveros-matt-taylor) [reply src/test/kotlin/coverosR3z/authentication/AuthenticationBDD.kt]: No, I will remove. Though we do need to puzzle out authUtilities logging as a TODO

File: src/test/kotlin/coverosR3z/authentication/AuthenticationBDD.kt (hunk 7 above)
Comment (coveros-matt-taylor) [reply src/test/kotlin/coverosR3z/authentication/AuthenticationBDD.kt:46]: I am not sure what _specifically_ this comment is referring to

File: src/test/kotlin/coverosR3z/authentication/AuthenticationBDD.kt (hunk 5 above)
Comment (coveros-matt-taylor) [reply src/test/kotlin/coverosR3z/authentication/AuthenticationBDD.kt:34]: I've gone with what I described, it just required me to move this Assertion Error to a slightly different place

Comment (7ep) [reply src/test/kotlin/coverosR3z/authentication/AuthenticationBDD.kt:34]: oh right.  Well, there's the other way - you log including the user where you can, and not where you can't.  You could have two calls like this:
```
class Logger(val cua : ICurrentUserAccessor) {
    fun info(msg: String) {
        println("INFO: ${cua.get().name}: $msg")
    }
}
fun logInfo(msg: String) {
    println("INFO: $msg")
}
```

Comment (7ep) [reply src/test/kotlin/coverosR3z/authentication/AuthenticationBDD.kt:34]: But maybe we can just leave it for later.

File: src/test/kotlin/coverosR3z/authentication/AuthenticationBDD.kt (hunk 7 above)
Comment (7ep) [reply src/test/kotlin/coverosR3z/authentication/AuthenticationBDD.kt:46]: This test has a lot of lines

**Review by 7ep — APPROVED** [review]

**Pull request merged**